src/main/java/com/example/petbuddybackend/service/care/state/TransitionManager.java

```diff
@@ -29,25 +34,43 @@ public void addTransition(
             BiConsumer<Care, CareStatus> onSuccess,
             Predicate<Care> prerequisite
     ) {
-        transitions.put(new Transition(role, fromStatus, toStatus), new TransitionAction(prerequisite, onSuccess));
+        transitions.put(new RoleTransition(role, fromStatus, toStatus), new TransitionAction(prerequisite, onSuccess));
     }
 
-    public Care transition(Care care, Role role, CareStatus newState) {
+    /**
+     * Transition state related to role
+     * */
+    public Care transition(Role role, Care care, CareStatus newState) {
         Transition transition = role.equals(Role.CARETAKER) ?
-                new Transition(role, care.getCaretakerStatus(), newState) :
-                new Transition(role, care.getClientStatus(), newState);
+                new RoleTransition(role, care.getCaretakerStatus(), newState) :
+                new RoleTransition(role, care.getClientStatus(), newState);
+
+        return performTransitions(care, List.of(transition));
+    }
+
+    /**
+     * Transition for both roles
+     * */
+    public Care transitionBothRoles(Care care, CareStatus newState) {
+        Transition clientTransition = new Transition(care.getClientStatus(), newState);
+        Transition caretakerTransition = new Transition(care.getCaretakerStatus(), newState);
+        return performTransitions(care, List.of(clientTransition, caretakerTransition));
+    }
 
-        if(!transitions.containsKey(transition)) {
-            throw new StateTransitionException(transition);
+    private Care performTransitions(Care care, List<Transition> transitions) {
+        if(transitions.stream().anyMatch(t -> !this.transitions.containsKey(t))) {
+            throw new StateTransitionException(transitions);
         }
 
-        TransitionAction transitionAction = transitions.get(transition);
+        List<TransitionAction> transitionActions = transitions.stream()
+                .map(this.transitions::get)
+                .toList();
 
-        if(!transitionAction.getPrerequisite().test(care)) {
-            throw new StateTransitionException(transition);
+        if(transitionActions.stream().anyMatch(t -> !t.getPrerequisite().test(care))) {
+            throw new StateTransitionException(transitions);
         }
 
-        transitionAction.getOnSuccess().accept(care, newState);
+        transitionActions.forEach(t -> t.getOnSuccess().accept(care, transitions.get(0).getToStatus()));
```

Comment:
raczej bym bral ToStatus od kazdej tranzycji, dla ktorej jest wykonywane to, bo moze byc tak ze przyjda 2 tranzycje z innym docelowym statusem. U nas to niemozliwe, ale dla czystosci samej metody bym tak zrobil

```suggestion
    private Care performTransitions(Care care, List<Transition> transitions) {
        if(transitions.stream().anyMatch(t -> !this.transitions.containsKey(t))) {
            throw new StateTransitionException(transitions);
        }

        Map<TransitionAction, CareStatus> transitionActions = transitions.stream()
                .collect(
                        HashMap::new,
                        (map, transition) -> map.put(this.transitions.get(transition), transition.getToStatus()),
                        HashMap::putAll
                );

        if(transitionActions.keySet().stream().anyMatch(t -> !t.getPrerequisite().test(care))) {
            throw new StateTransitionException(transitions);
        }
        
        transitionActions.forEach((transitionAction, newStatus) -> transitionAction.getOnSuccess().accept(care, newStatus));
        return care;
    }
```
cos takiego

Reply:
Faktycznie można coś z duplikatami pokombinować

Reply:
Minimalnie to zmieniłem żeby jako klucz był `Transition` bo doczytałem że może być problem z `hashCode` dla `Predicate` i `BiConsumer` jeżeli to się przekazuje przez lambde. Ale idea została taka sama